### PR TITLE
feat(Turborepo): add filesystem cookies on top of filewatching

### PR DESCRIPTION
**Pull request by gsoltis**
### Description

 - ports `cookie.go`
 - implements waiting for a roundtrip through the filesystem so filewatching queries can get consistent answers

### Testing Instructions

 - ports tests from `cookie_test.go`

Closes TURBO-1291

### CODE REVIEW

File: crates/turborepo-filewatch/src/cookie_jar.rs [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    fs::OpenOptions,
+    path::PathBuf,
+    sync::{atomic::AtomicUsize, Arc},
+    time::Duration,
+};
+
+use notify::{Event, EventKind};
+use thiserror::Error;
+use tokio::{
+    sync::{broadcast, oneshot, Mutex},
```
Comment (chris-olszewski) [inline crates/turborepo-filewatch/src/cookie_jar.rs:12]: I think we should be using the standard mutex here.

[tokio mutex vs std mutex](https://tokio.rs/tokio/tutorial/shared-state#on-using-stdsyncmutex) and the [alternatives suggested](https://tokio.rs/tokio/tutorial/shared-state#holding-a-mutexguard-across-an-await)

TL;DR use std unless you need to hold the mutex across an await and even then, you should try to avoid doing that

File: crates/turborepo-filewatch/src/cookie_jar.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+// result flattening is an unstable feature, so add a manual helper to do so.
+// This version is for unwrapping events coming from filewatching
+fn unwrap_event(
+    event: Result<Result<Event, NotifyError>, broadcast::error::RecvError>,
+) -> Result<Event, WatchError> {
+    match event {
+        Ok(event) => event.map_err(|e| e.into()),
+        Err(e) => Err(e.into()),
+    }
```
Comment (chris-olszewski) [inline crates/turborepo-filewatch/src/cookie_jar.rs:179]:
```suggestion
    Ok(event??)
```

File: crates/turborepo-filewatch/src/cookie_jar.rs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                    }
+                }
+            }
+        }
+    }
+}
+
+// result flattening is an unstable feature, so add a manual helper to do so.
+// This version is for unwrapping events coming from filewatching
+fn unwrap_event(
```
Comment (chris-olszewski) [inline crates/turborepo-filewatch/src/cookie_jar.rs:173]: Nit: I think `flatten_event` would be a lot better naming wise. `unwrap` make me think it'll panic if we encounter something that's not an event

File: crates/turborepo-filewatch/src/cookie_jar.rs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                .insert(cookie_path.as_std_path().to_owned(), tx);
+        }
+        let mut opts = OpenOptions::new();
+        opts.truncate(true).create(true).write(true);
+        {
+            // dropping the resulting file closes the handle
+            _ = cookie_path.open_with_options(opts)?;
+        }
+        let resp = tokio::time::timeout(self.timeout, rx).await?;
+        unwrap_resp(resp)
```
Comment (chris-olszewski) [inline crates/turborepo-filewatch/src/cookie_jar.rs:107]:
```suggestion
        Ok(resp??)
```

File: crates/turborepo-filewatch/src/cookie_jar.rs [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    match event {
+        Ok(event) => event.map_err(|e| e.into()),
+        Err(e) => Err(e.into()),
+    }
+}
+
+// result flattening is an unstable feature, so add a manual helper to do so.
+// This version is for flattening notifications coming back from the cookie
+// watching task
+fn unwrap_resp(
```
Comment (chris-olszewski) [inline crates/turborepo-filewatch/src/cookie_jar.rs:185]: This can get dropped if we make use of `??`

**Review by chris-olszewski — COMMENTED** [review]
First pass and it looks solid. I want to give the tests another look over with fresh eyes.

File: crates/turborepo-filewatch/src/cookie_jar.rs (hunk 1 above)
Comment (gsoltis) [reply crates/turborepo-filewatch/src/cookie_jar.rs:12]: Good reference. Updated to `std::sync::Mutex`

File: crates/turborepo-filewatch/src/cookie_jar.rs (hunk 3 above)
Comment (gsoltis) [reply crates/turborepo-filewatch/src/cookie_jar.rs]: `flatten_event` is definitely better, switched

File: crates/turborepo-filewatch/src/cookie_jar.rs (hunk 5 above)
Comment (gsoltis) [reply crates/turborepo-filewatch/src/cookie_jar.rs]: Dropped.

**Review by chris-olszewski — APPROVED** [review]
